rts/lra/lra-service-base/src/main/java/io/narayana/lra/LRAHttpClient.java

```diff
@@ -0,0 +1,64 @@
+package io.narayana.lra;
+
+import org.apache.http.client.HttpClient;
+import org.apache.http.client.ResponseHandler;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.FutureRequestExecutionService;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.impl.client.HttpRequestFutureTask;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class LRAHttpClient {
+    public static final long PARTICIPANT_TIMEOUT = 1; // number of seconds to wait for requests
+
+    private static LRAHttpClient client = new LRAHttpClient();
+
+    private FutureRequestExecutionService futureRequestExecutionService;
+
+    private LRAHttpClient() {
+        HttpClient httpClient = HttpClientBuilder.create().setMaxConnPerRoute(5).build();
+        ExecutorService executorService = Executors.newFixedThreadPool(5);
+
+        futureRequestExecutionService = new FutureRequestExecutionService(httpClient, executorService);
+    }
+
+    public static LRAHttpClient getClient() {
+        return client;
+    }
+
+    public ResponseHolder asyncRequest(HttpRequestBase httpMethod, long timelimit, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+        HttpRequestFutureTask<ResponseHolder> futureTask = submitFutureTask(httpMethod);
+
+        return futureTask.get(timelimit, unit);
```

Comment:
this seems to be a blocking call, maybe you can rename the method to just `request`? This seems confusing from the external point of view.

Reply:
Interesting observation, I expect the request itself is async. requestWithTimeout might be a more accurate name? Although I imagine the request itself is not cancelled so the request will be served at some point on the server side.

@mmusgrov what assumptions are made when the timelimit is expired but the request has not returned yet?

Reply:
Okay

Reply:
@tomjenkinson the request may well be still in progress but the caller is responsible for retrying the request - we make no assumptions about once only semantics so this is not an issue.

Reply:
Thanks